### PR TITLE
Integrate GolangCI-Lint

### PR DESCRIPTION
**Pull request by akwanmaroso**
#### Summary
Integrate GolangCI-Lint and refactor some code

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-plugin-nps/issues/76

### CODE REVIEW

**Comment by akwanmaroso** [comment]
I already signed it

**Comment by jasonblais** [comment]
Thanks @akwanmaroso!

**Comment by jasonblais** [comment]
@hanzei and @hmhealey this is ready for reviews (I don't permission to add you as reviewers)

File: build/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,5 @@
+FROM golang:1.15
```
Comment (hanzei) [inline build/Dockerfile:1]: These changes seem unrelated. Where are they from?

File: server/api.go [hunk 2]
```diff
@@ -148,13 +150,14 @@ func (p *Plugin) submitScore(w http.ResponseWriter, r *http.Request) {
 	}
 
 	var score int
-	if i, err := getScore(surveyResponse.Context["selected_option"].(string)); err != nil {
+	var i int64
+	var err error
+	if i, err = getScore(surveyResponse.Context["selected_option"].(string)); err != nil {
```
Comment (hanzei) [inline server/api.go:155]: Does moving the `getScore` call out of the `if` also fixes the linter error?

File: server/api.go [hunk 3]
```diff
@@ -169,7 +172,7 @@ func (p *Plugin) submitScore(w http.ResponseWriter, r *http.Request) {
 
 	// Thank the user for their feedback when they first answer the survey
 	if isFirstResponse {
-		p.CreateBotDMPost(userID, p.buildFeedbackRequestPost())
+		_, _ = p.CreateBotDMPost(userID, p.buildFeedbackRequestPost())
```
Comment (hanzei) [inline server/api.go:175]: Please log the error

File: server/survey.go [hunk 4]
```diff
@@ -82,22 +84,26 @@ func (p *Plugin) checkForNextSurvey(now time.Time) bool {
...
 		p.API.LogError("Failed to send notification of next survey to admins", "err", err)
 		return false
-	} else if !sent {
+	}
+
+	switch sent {
+	case false:
 		p.API.LogInfo("Not sending notification of next survey to admins since they already received one recently")
-	} else {
+	default:
```
Comment (hanzei) [inline server/survey.go:106]: Is this change really needed?

File: server/utils.go [hunk 5]
```diff
@@ -74,9 +72,9 @@ func (p *Plugin) IsBotDMChannel(channel *model.Channel) bool {
...
-	r := rand.New(rand.NewSource(p.now().UnixNano()))
-	delay := time.Duration(r.Int63n(int64(maxDelay) + 1))
+// func (p *Plugin) sleepUpTo(maxDelay time.Duration) {
+// 	r := rand.New(rand.NewSource(p.now().UnixNano()))
+// 	delay := time.Duration(r.Int63n(int64(maxDelay) + 1))
 
-	time.Sleep(delay)
-}
+// 	time.Sleep(delay)
+// }
```
Comment (hanzei) [inline server/utils.go:80]: Any objections on just removing this code?

**Review by hanzei — CHANGES_REQUESTED** [review]
Thanks for fixing all of these errors :+1:

File: build/Dockerfile (hunk 1 above)
Comment (akwanmaroso) [reply build/Dockerfile:1]: I am sorry for this one, I just follow the step from the issue so copy the Dockerfile but I forget to delete that, I fill fix that

File: server/api.go (hunk 2 above)
Comment (akwanmaroso) [reply server/api.go:155]: yes, because I the linter say the errors *if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)*.

File: server/api.go (hunk 3 above)
Comment (akwanmaroso) [reply server/api.go:175]: okey, i will add that

File: server/utils.go (hunk 5 above)
Comment (akwanmaroso) [reply server/utils.go:80]: Nothing, I ran the test, everything looks fine.

File: server/survey.go (hunk 4 above)
Comment (akwanmaroso) [reply server/survey.go:106]: I'm not sure about this, but when running golangci-lint, linter say if the *if-else chain* must change with *switch case*

File: server/api.go (hunk 3 above)
Comment (akwanmaroso) [reply server/api.go]: i fix this in this [4a8672](https://github.com/mattermost/mattermost-plugin-nps/pull/79/commits/4a8672f349b2bc34a822d17e42dfc25ace0212ed)

File: server/api.go (hunk 2 above)
Comment (hanzei) [reply server/api.go]: Given that you already dropped the previous `else`, could you double check if that is still the case?

File: server/api.go [hunk 6]
```diff
@@ -172,7 +172,11 @@ func (p *Plugin) submitScore(w http.ResponseWriter, r *http.Request) {
 
 	// Thank the user for their feedback when they first answer the survey
 	if isFirstResponse {
-		_, _ = p.CreateBotDMPost(userID, p.buildFeedbackRequestPost())
+		_, err := p.CreateBotDMPost(userID, p.buildFeedbackRequestPost())
+		if err != nil {
+			p.API.LogError("Failed to response feedback user")
+			w.WriteHeader(http.StatusInternalServerError)
```
Comment (hanzei) [inline server/api.go:178]: Let's just just an error here but don't write the header. The method should still continue and we don't want to `return` here.

File: server/api.go (hunk 2 above)
Comment (akwanmaroso) [reply server/api.go]: yes, and the result it's work and all linter is pass

**Review by hanzei — APPROVED** [review]
LGMT :+1:

File: server/survey.go [hunk 7]
```diff
@@ -82,48 +84,51 @@ func (p *Plugin) checkForNextSurvey(now time.Time) bool {
...
-	} else if !sent {
+	}
+
+	if !sent {
 		p.API.LogInfo("Not sending notification of next survey to admins since they already received one recently")
-	} else {
-		p.API.LogInfo("Sent notification of next survey to admins")
 	}
 
+	p.API.LogInfo("Sent notification of next survey to admins")
```
Comment (hmhealey) [inline server/survey.go:107]: This should go back into the else block since there's no return in the if statement

File: server/utils.go [hunk 8]
```diff
@@ -74,9 +72,9 @@ func (p *Plugin) IsBotDMChannel(channel *model.Channel) bool {
 	return true
 }
 
-func (p *Plugin) sleepUpTo(maxDelay time.Duration) {
-	r := rand.New(rand.NewSource(p.now().UnixNano()))
-	delay := time.Duration(r.Int63n(int64(maxDelay) + 1))
+// func (p *Plugin) sleepUpTo(maxDelay time.Duration) {
```
Comment (hmhealey) [inline server/utils.go:75]: If you could remove this, that'd be great. This was left over from some extremely unreliable code that never made it into production.

**Review by hmhealey — CHANGES_REQUESTED** [review]

**Comment by hmhealey** [comment]
Sorry for the delay in reviewing this. I thought I had submitted my review of this before the weekend, but I clearly didn't. I'll make sure to follow up with it more closely.

File: server/survey.go (hunk 7 above)
Comment (akwanmaroso) [reply server/survey.go]: okey

File: server/utils.go (hunk 8 above)
Comment (akwanmaroso) [reply server/utils.go]: i will remove that

**Comment by akwanmaroso** [comment]
no problem

**Review by hmhealey — APPROVED** [review]
Thanks! Looks good to me now

**Comment by hmhealey** [comment]
@DHaussermann For QA review, there shouldn't be anything changing here, so I think a quick smoke test should be good enough. Alternatively, if we need a qa build, we can merge it in and generate a build from the branch

**Comment by jasonblais** [comment]
@hmhealey @hanzei @jfrerich can you help add `hacktoberfest-accepted` label to this PR? (I don't have permissions to)

**Comment by hanzei** [comment]
I did a build test and a smoke test. No issues found.
